pilot/pkg/xds/eds.go

```diff
@@ -165,9 +167,35 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
 	// For existing endpoints, we need to do full push if service accounts change.
 	if saUpdated {
 		log.Infof("Full push, service accounts changed, %v", hostname)
-		fullPush = true
+		return true, true
+	}
+	if created {
+		return true, true
+	}
+	// Check if new Endpoints are ready to be pushed. This check
+	// will ensure that if a new pod comes with a non ready endpoint,
+	// we do not unnecessarily push that config to Envoy.
+	emap := make(map[string]*model.IstioEndpoint, len(oldIstioEndpoints))
+	for _, oie := range oldIstioEndpoints {
+		emap[oie.Address] = oie
 	}
-	return fullPush
+
+	for _, nie := range istioEndpoints {
+		if oie, exists := emap[nie.Address]; exists {
+			// If endpoint exists already, we should push if it's health status changes.
+			needPush = oie.HealthStatus != nie.HealthStatus
+		} else {
+			// If the endpoint does not exist in shards that means it is a
+			// new endpoint. Only send if it is healthy to avoid pushing endpoints
+			// that are not ready to start with.
+			needPush = nie.HealthStatus == model.Healthy
+		}
+		if needPush {
+			break
+		}
+	}
```

Comment:
Concerned with the performance, also should check the env flag

Reply:
What is the concern  - it is not n^2 any more and breaks at the earliest opportunity? Any suggestions?